### PR TITLE
Support Swift 4.2 Hashable

### PR DESCRIPTION
**Pull request by colemancda**
Updates implementation of `Hashable` in Foundation types to properly support Swift 4.2 and 5.

### CODE REVIEW

**Comment by colemancda** [comment]
@spevans Can you test this please?

File: Foundation/Date.swift [hunk 1]
```diff
@@ -137,9 +137,15 @@ public struct Date : ReferenceConvertible, Comparable, Equatable {
      */
     public static let distantPast = Date(timeIntervalSinceReferenceDate: -63114076800.0)
     
+    #if swift(>=4.2)
```
Comment (spevans) [inline Foundation/Date.swift]: The `swift` version check should not be needed as there is a branch per version.

File: Foundation/Measurement.swift [hunk 2]
```diff
@@ -35,10 +35,17 @@ public struct Measurement<UnitType : Unit> : ReferenceConvertible, Comparable, E
         self.value = value
         self.unit = unit
     }
-
+    
+    #if swift(>=4.2)
+    public func hash(into hasher: inout Hasher) {
+        unit.hash(into: &hasher)
+        value.hash(into: &hasher)
+    }
```
Comment (spevans) [inline Foundation/Measurement.swift]: I dont see this method on Darwin's Foundation (unless its in the latest beta?)

**Comment by spevans** [comment]
Could you add a testcase just calling the method for each of the types are modifying please

File: Foundation/Measurement.swift (hunk 2 above)
Comment (colemancda) [reply Foundation/Measurement.swift]: If you don't implement `func hash(into hasher: inout Hasher)` it uses `hashValue` as a fallback implementation, which on Darwin is just the `NSObject` hash.

**Comment by millenomi** [comment]
cc @lorentey who was already working on this, or similar.

**Comment by lorentey** [comment]
I suggest to hold this; this is somewhat trickier than it looks, and it has to be synced with the Foundation overlay. In https://github.com/apple/swift/pull/23832, I submitted a draft PR for the overlay parts; I'm preparing a similar PR for corelibs-foundation, too.

**Comment by lorentey** [comment]
Update: My corelibs-foundation PR is up at https://github.com/apple/swift-corelibs-foundation/pull/2118.

**Comment by millenomi** [comment]
@colemancda thank you for the contribution — I'm gonna close this because it has been superseded by #2118.

**Comment by colemancda** [comment]
Ok, if this is done better somewhere else, that's great.

**Pull request closed without merging**